manifest/provider/datasource.go

```diff
@@ -0,0 +1,164 @@
+	s.logger.Trace("[ReadDataSource][Request]\n%s\n", dump(*req))
+
+	resp := &tfprotov5.ReadDataSourceResponse{}
+
+	execDiag := s.canExecute()
+	if len(execDiag) > 0 {
+		resp.Diagnostics = append(resp.Diagnostics, execDiag...)
+		return resp, nil
+	}
+
+	rt, err := GetDataSourceType(req.TypeName)
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to determine data source type",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+
+	config, err := req.Config.Unmarshal(rt)
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to unmarshal data source configuration",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+
+	var dsConfig map[string]tftypes.Value
+	err = config.As(&dsConfig)
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to extract attributes from data source configuration",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+
+	rm, err := s.getRestMapper()
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "Failed to get RESTMapper client",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+	client, err := s.getDynamicClient()
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  "failed to get Dynamic client",
+			Detail:   err.Error(),
+		})
+		return resp, nil
+	}
+
+	var apiVersion, kind string
+	dsConfig["apiVersion"].As(&apiVersion)
+	dsConfig["kind"].As(&kind)
+
+	gvr, err := getGVR(apiVersion, kind, rm)
+	if err != nil {
+		return resp, err
+	}
+	gvk := gvr.GroupVersion().WithKind(kind)
+	ns, err := IsResourceNamespaced(gvk, rm)
+	if err != nil {
+		return resp, err
+	}
+	rcl := client.Resource(gvr)
+
+	var metadata map[string]tftypes.Value
+	dsConfig["metadata"].As(&metadata)
+
+	var name string
+	metadata["name"].As(&name)
+
+	var res *unstructured.Unstructured
+	if ns {
+		var namespace string
+		metadata["namespace"].As(&namespace)
+		res, err = rcl.Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
+	} else {
+		res, err = rcl.Get(ctx, name, metav1.GetOptions{})
+	}
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return resp, nil
+		}
+		d := tfprotov5.Diagnostic{
+			Severity: tfprotov5.DiagnosticSeverityError,
+			Summary:  fmt.Sprintf("Failed to get data source"),
+			Detail:   err.Error(),
+		}
+		resp.Diagnostics = append(resp.Diagnostics, &d)
+		return resp, nil
+	}
+
+	objectType, err := s.TFTypeFromOpenAPI(ctx, gvk, false)
```

Comment:
A point could be made that this should best happen before the actual API Get call. Reason being that it might serve as validation for the correctness of the input attributes independently of potential permission errors on the resource Get call.
What I'm not sure about is whether client-go distinguishes (with different error codes) between a RBAC permission issue on a valid path and an invalid path (likely also reported by 'not found').